File: Algorithm_problem_solving/Baek-joon/1644/1644.py
# https://www.acmicpc.net/problem/1644
import sys

input = sys.stdin.readline


def eratos(n: int):
    """
    에라토스 테네스의 체로 소수인 것에면 True체크를 해서 리턴한다.

    Args:
        n (int): 소수가 들어있는 최대의 범위 지정

    Returns:
        int: 소수가 표시된 배열을 반환한다.
    """
    seive = [True] * (n + 2)
    sqrtN = int(n ** 0.5)
    seive[0] = False
    seive[1] = False
    for i in range(2, sqrtN + 1):
        if seive[i]:
            for j in range(i * i, n + 1, i):
                seive[j] = False
    return seive


def serial_prime_sum(seive:list, n:int):
    """
    연속된 소수의 합으로 n이 되는 경우의 수를 반환한다.
    Args:
        seive (list): 에라토스테네스의 체로 걸러진 소수의 집합
        n (int): 원하는 수 n

    Returns:
        int: n이 되는 경우의 수
    """
    answer = 0      # 반환할 값
    numList = []    # 현재 연속된 수들의 집합
    num, numSum = 1, 0  # 현재 체크하고 있는 수, 현재 연속된 수들의 합
    while num <= n + 1:
        # 연속된 소수의 합이 n보다 커지면 앞에서 빼고, n보다 작아지면 뒤에서 붙힌다.
        if numSum == n:
            answer += 1
        if numSum >= n:
            numSum -= numList[0]
            numList.pop(0)
        else:
            while not seive[num]:
                num += 1
            numSum += num
            numList.append(num)
            num += 1
    return answer


if __name__ == "__main__":
    n = int(input().strip())
    seive = eratos(n)

    print(serial_prime_sum(seive, n))

File: Algorithm_problem_solving/Baek-joon/1644/1644.kt
// https://www.acmicpc.net/problem/1644

/**
 * 에라토스 테네스의 체로 소수들을 찾는다.
 *
 * @param n n의 범위 까지의 소수
 * @return 소수인 수는 true로 체크되어있는 n까지의 배열
 */
fun eratostenes(
    num: Int
): Array<Boolean> {
    val sieve = Array(num + 1) { true }
    sieve[0] = false
    sieve[1] = false

    for (i in 2..Math.sqrt(num.toDouble()).toInt()) {
        if (sieve[i]) {
            for (j in i * i..num step i) {
                sieve[j] = false
            }
        }
    }

    return sieve
}

/**
 * 연속된 소수의 합이 goal이 되는 경우의 수를 찾는다
 *
 * @param goal 목표 합계 값
 * @param sieve 소수 여부를 나타내는 배열
 * @return 연속된 소수의 합으로 goal이 만들어지는 경우의 수
 */
fun getPrimeSum(
    goal: Int,
    seive: Array<Boolean>
): Int {
    var answer = 0
    val primeList: MutableList<Int> = mutableListOf()

    var pivot = 2
    var primeSum = 0

    while (pivot <= goal + 1) {
        if (primeSum == goal) {
            answer += 1
        }
        if (primeSum <= goal) {
            while (pivot <= goal && !seive[pivot]) {
                pivot++
            }
            primeSum += pivot
            primeList.add(pivot++)
        } else {
            primeSum -= primeList.first()
            primeList.removeAt(0)
        }
    }

    return answer
}

fun main() {
    val num = readln().toInt()
    val eratosSeive = eratostenes(num)

    println(getPrimeSum(num, eratosSeive))
}